lightning/src/util/ser.rs

```diff
@@ -913,6 +915,82 @@ impl Readable for String {
 	}
 }
 
+/// Represents a hostname with validation for serialization purposes.
+/// The hostname follows the preferred form summarized in e.g. RFC 3696.
+/// Its length is guaranteed to be representable by a single byte.
+/// This serialization is used by Bolt 7 hostnames.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Hostname(String);
+impl Hostname {
+	/// Returns the length of the hostname.
+	pub fn len(&self) -> u8 {
+		(&self.0).len() as u8
+	}
+}
+impl Deref for Hostname {
+	type Target = String;
+
+	fn deref(&self) -> &Self::Target {
+		&self.0
+	}
+}
+impl From<Hostname> for String {
+	fn from(short_s: Hostname) -> Self {
+		short_s.0
+	}
+}
+impl TryFrom<String> for Hostname {
+	type Error = ();
+
+	fn try_from(s: String) -> Result<Self, Self::Error> {
+		if s.len() > 255 {
+			return Err(());
+		}
+		let mut labels = s.split('.').enumerate().peekable();
+		while let Some((label_idx, label)) = labels.next() {
+			let last_label = labels.peek().is_none();
+			let bytes = label.as_bytes();
+			// Trailing period for fully-qualified name.
+			if bytes.len() == 0 {
+				if !last_label || label_idx == 0 {
+					return Err(());
+				}
+			}
+			if bytes.len() > 63 {
+				return Err(());
+			}
+			for byte_idx in 0 .. bytes.len() {
+				let b = bytes[byte_idx];
+				if !b.is_ascii_alphanumeric() && b != b'-' {
+					return Err(());
+				}
+				if b == b'-' && (byte_idx == 0 || byte_idx == bytes.len() - 1) {
+					return Err(());
+				}
+			}
+		};
+		Ok(Hostname(s))
+	}
+}
+impl Writeable for Hostname {
+	#[inline]
+	fn write<W: Writer>(&self, w: &mut W) -> Result<(), io::Error> {
+		self.len().write(w)?;
+		w.write_all(self.as_bytes())
+	}
+}
+impl Readable for Hostname {
+	#[inline]
+	fn read<R: Read>(r: &mut R) -> Result<Hostname, DecodeError> {
+		let len: u8 = Readable::read(r)?;
+		let mut vec = Vec::with_capacity(len as usize);
+		vec.resize(len as usize, 0);
+		r.read_exact(&mut vec)?;
+		let s = String::from_utf8(vec).map_err(|_| DecodeError::InvalidValue)?;
```

Comment:
Done. I think it is still useful to still have `Hostname` wrap a `String` for dereferencing purposes, but now it is also possible to avoid duplicate UTF-8 validation.

Reply:
Alright, leaving it a String is fine I guess.